stack/trigger/trigger.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package trigger provides functionality that help manipulate stacks triggers.
+package trigger
+
+import (
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/project"
+	"github.com/rs/zerolog/log"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Info represents the parsed contents of a trigger
+// for triggers created by Terramate.
+type Info struct {
+	// Ctime is unix timestamp of when the trigger was created.
+	Ctime int64
+	// Reason is the reason why the trigger was created, if any.
+	Reason string
+}
+
+const triggersDir = ".tmtriggers"
+
+// StackPath accepts a trigger file path and returns the path of the stack
+// that is triggered by the given file. If the given file is not a stack trigger
+// at all it will return false.
+func StackPath(triggerFile project.Path) (project.Path, bool) {
+	const triggersPrefix = "/" + triggersDir
+
+	if !strings.HasPrefix(triggerFile.String(), triggersPrefix) {
+		return project.Path("/"), false
+	}
+
+	stackPath := strings.TrimPrefix(triggerFile.String(), triggersPrefix)
+	stackPath = path.Dir(stackPath)
+	return project.Path(stackPath), true
+}
+
+// ParseFile will parse the given trigger file.
+func ParseFile(path string) (Info, error) {
+	parser := hclparse.NewParser()
+	parsed, diags := parser.ParseHCLFile(path)
+	if diags.HasErrors() {
+		return Info{}, errors.E(diags, "parsing trigger file")
+	}
+	rootContent, diags := parsed.Body.Content(&hcl.BodySchema{
+		Blocks: []hcl.BlockHeaderSchema{
+			{
+				Type: "trigger",
+			},
+		},
+	})
+	if diags.HasErrors() {
+		return Info{}, errors.E(diags, "checking root schema")
+	}
+
+	triggerBlock := rootContent.Blocks[0]
+	triggerContent, diags := triggerBlock.Body.Content(&hcl.BodySchema{
+		Attributes: []hcl.AttributeSchema{
+			{
+				Name:     "ctime",
+				Required: true,
+			},
+			{
+				Name:     "reason",
+				Required: true,
+			},
+		},
+	})
+	if diags.HasErrors() {
+		return Info{}, errors.E(diags, "checking trigger schema")
+	}
+
+	errs := errors.L()
+	info := Info{}
+
+	for _, attribute := range triggerContent.Attributes {
```

Comment:
I think it'd be better if we range over a sorted list of attrs as we do in other packages to avoid randomization in the error list.

Reply:
Nice catch :100:, done